fili-core/src/test/groovy/com/yahoo/bard/webservice/async/AsyncResultsNotReadySpec.groovy

```diff
@@ -0,0 +1,71 @@
+package com.yahoo.bard.webservice.async
+
+import com.yahoo.bard.webservice.async.jobs.jobrows.DefaultJobField
+import com.yahoo.bard.webservice.async.jobs.jobrows.DefaultJobStatus
+
+import spock.lang.Timeout
+
+import javax.ws.rs.core.Response
+
+/**
+ * Verifies that if the result of an asynchronous query is not ready yet, then the user get back the exact same payload
+ * from the jobs endpoint they received from the data endpoint.
+ */
+@Timeout(30)
+class AsyncResultsNotReadySpec extends AsyncFunctionalSpec {
+
+    static final String QUERY =
+            "http://localhost:9998/data/shapes/day/color?dateTime=2016-08-30%2F2016-08-31&metrics=height&asyncAfter=0"
+
+    @Override
+    Map<String, Closure<String>> getResultsToTargetFunctions() {
+        return [
+                data: {"data/shapes/day/color"},
+                jobs: {AsyncTestUtils.buildTicketLookup(it.data.readEntity(String))},
+                results: {payloads ->
+                    String dataPayload = payloads.data.readEntity(String)
+                    "jobs/${AsyncTestUtils.getJobFieldValue(dataPayload, DefaultJobField.JOB_TICKET.name)}/results"
+                }
+        ]
+    }
+
+    @Override
+    Map<String, Closure<Void>> getResultAssertions() {
+        return [
+                data: {Response response ->
+                    assert response.status == 202
+                    AsyncTestUtils.validateJobPayload(response.readEntity(String), QUERY, DefaultJobStatus.PENDING.name)
+                },
+                jobs: {Response response ->
+                    assert response.status == 200
+                    AsyncTestUtils.validateJobPayload(response.readEntity(String), QUERY, DefaultJobStatus.PENDING.name)
+                },
+                results: {Response response ->
+                    assert response.status == 200
+                    AsyncTestUtils.validateJobPayload(response.readEntity(String), QUERY, DefaultJobStatus.PENDING.name)
+                }
+        ]
+    }
+
+    @Override
+    Map<String, Closure<Map<String, List<String>>>> getQueryParameters() {
+        [
+                data: {[
+                        metrics: ["height"],
+                        asyncAfter: ["0"],
+                        dateTime: ["2016-08-30/2016-08-31"]
+                ]},
+                jobs: {[:]},
+                results: { [ asyncAfter: ["0"] ] }
+        ]
+    }
+
+    @Override
+    Closure<String> getFakeDruidResponse() {
+        return {
+            //30 seconds, the same length as the test timeout. In this test, we don't want Druid to respond at all.
+            Thread.sleep(30000)
```

Comment:
A different (and _possibly_ better) way to do this would be to use a Latch (like a countdown latch), which can be shared between the thread the `getFakeDruidResponse` runs on and the thread that the request validation code runs on. Then, when the final assertion (or something) runs, it can "un-latch" and allow the Druid response to flow. That prevents the need to block that thread for 30 seconds, letting it be blocked only as long as needed.

This is the general mechanism for dealing with ordering across threads that is timing / load-safe, and doesn't waste any more time than needed, since it's not time-based.